geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/builder/FeatureTypeBuilder.scala

```diff
@@ -0,0 +1,119 @@
+package org.locationtech.geomesa.utils.geotools.builder
+
+import java.util.UUID
+
+import org.geotools.feature.`type`.AttributeDescriptorImpl
+import org.geotools.feature.{AttributeImpl, ComplexAttributeImpl, FeatureImpl, NameImpl}
+import org.geotools.filter.identity.FeatureIdImpl
+import org.geotools.gml3.GMLSchema
+import org.opengis.feature.`type`._
+import org.opengis.feature.{ComplexAttribute, Feature, Property}
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
+object FeatureTypeBuilder {
+  def namespace(name: String) = new TopLevelFeatureTypeBuilder().namespace(name)
+
+  def descriptor(namespace: String, local: String): DescriptorBuilder[_<:AttributeDescriptor] = {
+    new TopLevelDescriptorBuilder().namespace(namespace).name(local)
+  }
+
+  def feature(`type`: ComplexType, name: String, id: String, values: Any*): Feature = {
+    val desc: AttributeDescriptor = new AttributeDescriptorImpl(`type`, new NameImpl(`type`.getName.getNamespaceURI, name), 1, 1, false, null)
+    val props = for((t,v)<-`type`.getDescriptors.zip(values)) yield new AttributeImpl(v, t.asInstanceOf[AttributeDescriptor], null)
+    new FeatureImpl(props, desc, new FeatureIdImpl(id))
+  }
+}
+
+object FeatureBuilder {
+  def apply(t: ComplexType)(fn: ComplexAttributeBuilder=>ComplexAttributeBuilder) = fn(new ComplexAttributeBuilder(t, makeDescriptor(t))).end
+  def apply(t: FeatureType)(fn: FeatureBuilder=>FeatureBuilder):Feature = fn(new FeatureBuilder(t, makeDescriptor(t))).end
+
+  def makeDescriptor(t:ComplexType): AttributeDescriptor = new AttributeDescriptorImpl(t, t.getName, 1, 1, false, null)
+
+  def isDescendedFrom(child: PropertyType, ancestor: PropertyType): Boolean =
+    Iterator.iterate(child)(_.getSuper).takeWhile(_ != null).contains(ancestor)
+
+  def subclassByRestriction(child: PropertyType) =
+    isDescendedFrom(child, GMLSchema.FEATUREPROPERTYTYPE_TYPE)
+
+  def allProps(ct:ComplexType):Iterable[PropertyDescriptor] = ct.getSuper match {
+    case st:ComplexType if subclassByRestriction(st) => allProps(st).drop(ct.getDescriptors.size())++ct.getDescriptors

```

Comment:
It seems like the first several descriptors are replaced here.

Is it clear why the ones to replace are up first?